### PR TITLE
Re-infer nullable result of switch expression

### LINKED ISSUE #34233: Re-infer nullable result of switch expression
In the nullable walker, we need to recompute the result type and state of the switch expression based on the result type and state of all of the arms.  This can be done in a way similar to how it is done for an implicit array creation expression.

This issue link is being used in a code comment where it needs to be fixed.

### PR DESCRIPTION
**Pull request by gafter**
Fixes #34233

### CODE REVIEW

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs [hunk 1]
```diff
@@ -431,21 +431,63 @@ public override BoundNode VisitSwitchExpression(BoundSwitchExpression node)
...
+                conversions.Add(conversion);
+                resultTypes.Add(VisitRvalueWithState(expression));
                 Join(ref endState, ref this.State);
             }
 
+            // Build placeholders for inference in order to preserve annotations.
+            TypeSymbol inferredType = null;
+            var placeholderBuilder = ArrayBuilder<BoundExpression>.GetInstance(n);
+            for (int i = 0; i < n; i++)
+                placeholderBuilder.Add(CreatePlaceholderIfNecessary(expressions[i], resultTypes[i].ToTypeWithAnnotations()));
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs]: >placeholderBuilder.Add [](start = 16, length = 22)

Consider moving this into `foreach` above.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs [hunk 2]
```diff
@@ -431,21 +431,63 @@ public override BoundNode VisitSwitchExpression(BoundSwitchExpression node)
...
-                VisitRvalue(arm.Value);
+                (BoundExpression expression, Conversion conversion) = RemoveConversion(arm.Value, includeExplicitConversions: false);
+                expressions.Add(expression);
+                conversions.Add(conversion);
+                resultTypes.Add(VisitRvalueWithState(expression));
                 Join(ref endState, ref this.State);
             }
 
+            // Build placeholders for inference in order to preserve annotations.
+            TypeSymbol inferredType = null;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs]: >TypeSymbol inferredType = null; [](start = 12, length = 31)

Declaration can be moved to `= InferBestType(...);`.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs [hunk 3]
```diff
@@ -431,21 +431,63 @@ public override BoundNode VisitSwitchExpression(BoundSwitchExpression node)
...
+            var placeholderBuilder = ArrayBuilder<BoundExpression>.GetInstance(n);
+            for (int i = 0; i < n; i++)
+                placeholderBuilder.Add(CreatePlaceholderIfNecessary(expressions[i], resultTypes[i].ToTypeWithAnnotations()));
+            var placeholders = placeholderBuilder.ToImmutableAndFree();
+
+            HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+            inferredType = BestTypeInferrer.InferBestType(placeholders, _conversions, ref useSiteDiagnostics);
+            if (inferredType is null)
+                inferredType = node.Type.SetUnknownNullabilityForReferenceTypes();
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs]: `??`

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs [hunk 4]
```diff
@@ -431,21 +431,63 @@ public override BoundNode VisitSwitchExpression(BoundSwitchExpression node)
...
+
+            // Convert elements to best type to determine element top-level nullability and to report nested nullability warnings
+            var inferredState = NullableFlowState.NotNull;
+            for (int i = 0; i < n; i++)
+            {
+                var placeholder = placeholders[i];
+                var oneResultType = ApplyConversion(placeholder, placeholder, conversions[i], inferredTypeWithAnnotations, resultTypes[i], checkConversion: true,
+                    fromExplicitCast: false, useLegacyWarnings: false, AssignmentKind.Assignment, reportRemainingWarnings: true, reportTopLevelWarnings: false);
+                resultTypes[i] = oneResultType;
+                inferredState = inferredState.Join(oneResultType.State);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs]: >inferredState = inferredState.Join(oneResultType.State); [](start = 16, length = 56)

Should `inferredState` be using `BestTypeInferrer.GetNullableState()`?
```
var inferredState = BestTypeInferrer.GetNullableState(resultTypes);
```

**Comment by gafter** [comment]
@cston I believe I've responded to all of your comments.  Anything else?

**Comment by gafter** [comment]
@dotnet/roslyn-compiler May I please have a second review?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesVsPatterns.cs [hunk 5]
```diff
@@ -1834,5 +1834,89 @@ void M4((string? s1, string? s2)? n)
...
+        [Fact]
+        [WorkItem(34233, "https://github.com/dotnet/roslyn/issues/34233")]
+        public void SwitchExpressionResultType_01()
+        {
+            CSharpCompilation c = CreateNullableCompilation(@"
+class Test
+{
+    void Test1(int i, object? x, object y)
+    {
+        _ = i switch { 1 => x, _ => y }/*T:object?*/;
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesVsPatterns.cs]: Would it be worth it to add a test with a when clause affecting the state of `x`? For example, `_ = i switch { 1 when x != null => x, _ => y}/*T:object!*/;`

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs [hunk 6]
```diff
@@ -431,21 +431,60 @@ public override BoundNode VisitSwitchExpression(BoundSwitchExpression node)
                 ReportSafetyDiagnostic(ErrorCode.WRN_SwitchExpressionNotExhaustiveForNull, ((SwitchExpressionSyntax)node.Syntax).SwitchKeyword.GetLocation());
             }
 
+            // collect expressions, conversions and result types
+            int n = node.SwitchArms.Length;
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs]: >n [](start = 16, length = 1)

Consider using a more descriptive name, such as `numSwitchArms`

**Comment by gafter** [comment]
@333fred I think I've addressed your comments.  Do you have any others?

**Pull request merged**